app/service/learning_svc.py

```diff
@@ -35,43 +37,49 @@ async def build_model(self):
                         self.model.add(variables)
         self.model = set(self.model)
 
-    async def learn(self, facts, link, blob):
+    async def learn(self, facts, link, blob, operation=None):
         decoded_blob = b64decode(blob).decode('utf-8')
 
         found_facts = []
         for parser in self.parsers:
             try:
                 for fact in parser.parse(decoded_blob):
-                    await self._save_fact(link, facts, fact)
                     found_facts.append(fact)
             except Exception as e:
                 self.log.error(e)
-        await self._update_scores(link, facts, increment=len(found_facts))
-        await self._build_relationships(link, found_facts)
+        await _update_scores(operation=None, increment=len(found_facts), used=facts, facts=link.facts)
+        await self._store_results(link, found_facts, operation)
 
     """ PRIVATE """
 
     @staticmethod
-    async def _update_scores(link, facts, increment):
-        for uf in link.facts:
-            for found_fact in facts:
-                if found_fact.unique == uf.unique:
-                    found_fact.score += increment
-                    break
-
-    @staticmethod
-    async def _save_fact(link, facts, fact):
-        if all(fact.trait) and not any(f.trait == fact.trait and f.value == fact.value for f in facts):
+    async def _save_fact(link, facts, fact, operation=None):
+        if all(fact.trait) and not any(f.trait == fact.trait and f.value == fact.value and f.source == link.id
+                                       for f in facts):
             fact.collected_by = link.paw
             fact.technique_id = link.ability.technique_id
+            fact.links = [link]
+            fact.relationships = []
+            fact.source_type = OriginType.LEARNED.name
+            fact.source = operation.id if operation else link.id
```

Comment:
should this be above the conditional statement so that the differentiation of `operation.id` vs `link.id` is used in determining is the fact already exists?  As if it is during an operation, we will add the fact every time as we will check against `link.id` for checking existence and then store with the `operation.id`

Another benefit of this is that you can just use your equality comparison and do `any(fact == f for f in facts)` which will cover your three items.

Reply:
That's a fair point, good catch.